pkg/controller/pod/pod_controller.go

```diff
@@ -250,3 +264,106 @@ func (r *ReconcilePod) createK8sObject(obj *unstructured.Unstructured) error {
 	}
 	return nil
 }
+
+func (r *ReconcilePod) recreateNodeAgentPodsIfInvalidResolvConf(
+	resName string) error {
+	podsInCLB, err := identifyPodsInCLBDueToInvalidResolvConf(r.client)
+	if err != nil {
+		log.Error(err, "Could not identify if any pod is in CLB because "+
+			"of invalid resolv.conf")
+		return err
+	}
+	if len(podsInCLB) > 0 && !deletePods(podsInCLB, r.client) {
+		err := errors.New("Error occured while trying to restart pods in " +
+			"CLB because of invalid resolv.conf")
+		log.Error(err, "")
+		return err
+	}
+	return nil
+}
+
+func identifyPodsInCLBDueToInvalidResolvConf(c client.Client) (
+	[]corev1.Pod, error) {
+	var podsInCLB []corev1.Pod
+	podList := &corev1.PodList{}
+	nodeAgentLabelSelector := labels.SelectorFromSet(
+		map[string]string{"component": operatortypes.NsxNodeAgentDsName})
+	err := c.List(context.TODO(), podList, &client.ListOptions{
+		LabelSelector: nodeAgentLabelSelector})
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		if isNodeAgentContainerInCLB(&pod) {
+			nodeAgentLogs, err := getContainerLogsInPod(
+				&pod, operatortypes.NsxNodeAgentContainerName)
+			if err != nil {
+				log.Error(err, "Error occured while getting container logs")
+				return nil, err
+			}
+			if strings.Contains(
+				nodeAgentLogs, "Failed to connect k8s apiserver") {
+				log.Info(fmt.Sprintf(
+					"Pod %v is in CLB because of invalid resolv.conf. "+
+						"It shall be restarted", pod.Name))
```

Comment:
I might say the failure above does not guarantee the resolv.conf is invalid?

Reply:
since we are using logs, choose your weapon:
2020-09-24T09:14:57.985Z compute-dhcp-1.openshift1 NSX 21 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="ERROR" security="True" errorCode="NCP05001"] nsx_ujo.ncp.k8s.kubernetes k8s connection failed: Failed to connect k8s apiserver: HTTPSConnectionPool(host='api.upi.openshift.test', port=6443): Max retries exceeded with url: /api/v1/services/?labelSelector=dummy.key%3Ddummy.value (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f1d929a1710>: Failed to establish a new connection: [Errno -2] Name or service not known',))
2020-09-24T09:14:57.986Z compute-dhcp-1.openshift1 NSX 21 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="DEBUG" security="True"] nsx_ujo.agent.cni_watcher K8s configuration validation failed
2020-09-24T09:14:58.146Z compute-dhcp-1.openshift1 NSX 11 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="DEBUG"] oslo_privsep.comm EOF on privsep read channel
Traceback (most recent call last):
  File "/opt/rh/rh-python36/root/usr/bin/nsx_node_agent", line 10, in <module>
    sys.exit(main())
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/cmd/nsx_node_agent.py", line 19, in main
    agent_main.main()
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/common/privilege.py", line 35, in _wrap
    return self._keepalive_wrap(func, *args, **kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/common/privilege.py", line 48, in _keepalive_wrap
    return super(PrivContextPlus, self)._wrap(func, *args, **kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/priv_context.py", line 207, in _wrap
    return self.channel.remote_call(name, args, kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/daemon.py", line 192, in remote_call
    result = self.send_recv((Message.CALL.value, name, args, kwargs))
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/comm.py", line 165, in send_recv
    reply = future.result()
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/comm.py", line 108, in result
    raise self.error
OSError: Premature eof waiting for privileged process

Perhaps "Failed to establish a new connection"? Or a different mechanism to detect?

Reply:
observed this error from a newer image:
2020-10-06T22:03:52.680Z controle-plane-dhcp-1.occl443 NSX 22 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="ERROR" security="True" errorCode="NCP05001"] nsx_ujo.ncp.k8s.kubernetes k8s connection failed: Failed to connect k8s apiserver: HTTPSConnectionPool(host='api.upi.openshift.test', port=6443): Max retries exceeded with url: /api/v1/services/?labelSelector=dummy.key%3Ddummy.value (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f164c0a6d68>: Failed to establish a new connection: [Errno -2] Name or service not known',))
2020-10-06T22:03:52.680Z controle-plane-dhcp-1.occl443 NSX 22 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="DEBUG" security="True"] nsx_ujo.agent.cni_watcher K8s configuration validation failed
2020-10-06T22:03:52.680Z controle-plane-dhcp-1.occl443 NSX 22 - [nsx@6876 comp="nsx-container-node" subcomp="nsx_node_agent" level="CRITICAL"] concurrent.futures Exception in worker
Traceback (most recent call last):
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/thread.py", line 69, in _worker
    work_item.run()
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/thread.py", line 58, in run
    self.future.set_exception(exc)
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/_base.py", line 534, in set_exception
    self._invoke_callbacks()
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/_base.py", line 324, in _invoke_callbacks
    callback(self)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/daemon.py", line 471, in _call_back
    reply = result.result()
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/daemon.py", line 447, in _process_cmd
    ret = func(*f_args, **f_kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/common/privilege.py", line 44, in _wrap
    return self._keepalive_wrap(func, *args, **kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/common/privilege.py", line 55, in _keepalive_wrap
    return super(PrivContextPlus, self)._wrap(func, *args, **kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/oslo_privsep/priv_context.py", line 216, in _wrap
    return func(*args, **kwargs)
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/agent/main.py", line 19, in main
    agent.node_agent_run()
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/agent/agent.py", line 315, in node_agent_run
    cni_watcher_instance = cni_watcher.CNIWatcher()
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/agent/cni_watcher_lin.py", line 30, in __init__
    super(CNIWatcher, self).__init__()
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/nsx_ujo/agent/cni_watcher.py", line 40, in __init__
    sys.exit(1)

We could perhaps use 'Failed to establish a new connection: [Errno -2] Name or service not known' instead?

Reply:
seems better indeed